src/cudamatrix/cu-matrix.cc

```diff
@@ -1700,6 +1700,93 @@ double TraceMatMat(const CuMatrixBase<double> &A,
                    const CuMatrixBase<double> &B,
                    MatrixTransposeType trans);
 
+template<typename Real>
+void AddMatMatBatched(const Real alpha, std::vector<CuSubMatrix<Real>* > &C,
+		const std::vector<CuSubMatrix<Real>* > &A, MatrixTransposeType transA,
+		const std::vector<CuSubMatrix<Real>* > &B, MatrixTransposeType transB,
+		const Real beta) {
+    // the batch counts should be the same	
+    int32 batchCountA = static_cast<int32>(A.size()), batchCountB = static_cast<int32>(B.size()), batchCountC = static_cast<int32>(C.size());
+    KALDI_ASSERT((batchCountA == batchCountB) && (batchCountB == batchCountC));
+    int32 batchCount = batchCountA;
+
+    if (batchCount == 0) return;
+    
+    // all elements must have the same num-rows, num-cols and stride
+    for (int32 i = 0; i < batchCount-1; i++) {
+      KALDI_ASSERT(A[i]->NumRows() == A[i+1]->NumRows());
+      KALDI_ASSERT(A[i]->NumCols() == A[i+1]->NumCols());
+      KALDI_ASSERT(A[i]->Stride() == A[i+1]->Stride());
+      KALDI_ASSERT(B[i]->NumRows() == B[i+1]->NumRows());
+      KALDI_ASSERT(B[i]->NumCols() == B[i+1]->NumCols());
+      KALDI_ASSERT(B[i]->Stride() == B[i+1]->Stride());
+      KALDI_ASSERT(C[i]->NumRows() == C[i+1]->NumRows());
+      KALDI_ASSERT(C[i]->NumCols() == C[i+1]->NumCols());
+      KALDI_ASSERT(C[i]->Stride() == C[i+1]->Stride());
+    }
+    // CUBLAS is col-major, cudamatrix is row-major, how to do the mapping?
+    // keep trans..., just swap A&B matrices: A->B B->A
+    MatrixIndexT m = ((transB==kTrans)? B[0]->NumRows() : B[0]->NumCols()); 
+    MatrixIndexT n = ((transA==kTrans)? A[0]->NumCols() : A[0]->NumRows());
+    MatrixIndexT k = ((transB==kTrans)? B[0]->NumCols() : B[0]->NumRows());
+    MatrixIndexT k1 = ((transA==kTrans)? A[0]->NumRows() : A[0]->NumCols());
+
+    KALDI_ASSERT(m == C[0]->NumCols());
+    KALDI_ASSERT(n == C[0]->NumRows());
+    KALDI_ASSERT(k == k1);
+
+    if (m == 0) return;
+
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    // place all data (A[batchCount], B[batchCount], C[batchCount]) in memory together
+    // in order to make only one call of cudaMemcpy 	  
+    Real **DABC_arr = (Real**)CuDevice::Instantiate().Malloc(3*batchCount*sizeof(Real*));
+    Real **DA_arr = DABC_arr, **DB_arr = DABC_arr + batchCount, **DC_arr = DABC_arr + 2 * batchCount;
+    Real **HABC_arr = (Real**)malloc(3*batchCount*sizeof(Real*));
+    Real **HA_arr = HABC_arr, **HB_arr = HABC_arr + batchCount, **HC_arr = HABC_arr + 2 * batchCount; 
+    
+    for (int32 i = 0; i < batchCount; i++) {
+      memcpy(HA_arr+i, &(A[i]->data_), sizeof(Real*));
```

Comment:
Here, using memcpy is overkill.  Just do e.g.
Ha_arr[i] = A[i]->data_;
Also, although you'll still have to do a const_cast, it would be better to make DB_arr and DC_arr and HB_arr and HC_arr of type const Real**.  
When allocating HABC_arr, better to use the C++ new [] mechanism, e.g.
  Real **HABC_arr = new Real* [3*batchCount];
Also better to use lowercase variables as much as possible, and fewer abbreviations, per Google style guid e.g.  host_abc_array, and device_abc_array.